### PR TITLE
Graphml correctness witnesses

### PR DESCRIPTION
**Pull request by peterschrammel**
This covers the first 7 commits of https://github.com/diffblue/cbmc/pull/363

### CODE REVIEW

File: src/goto-programs/graphml_witness.cpp [hunk 1]
```diff
@@ -6,6 +6,10 @@ Author: Daniel Kroening
 
 \*******************************************************************/
 
+#include <iostream>
+
```
Comment (tautschnig) [inline src/goto-programs/graphml_witness.cpp]: We shouldn't need iostream

File: src/goto-programs/graphml_witness.cpp [hunk 2]
```diff
@@ -89,32 +93,55 @@ std::string graphml_witnesst::convert_assign_rec(
...
-         c_it->get_is_padding() ||
+      if(comp.type().id()==ID_code ||
+         comp.get_is_padding() ||
          // for some reason #is_padding gets lost in *some* cases
-         has_prefix(id2string(c_it->get_name()), "$pad"))
+         has_prefix(id2string(comp.get_name()), "$pad"))
       {
-        ++c_it;
         continue;
       }
```
Comment (tautschnig) [inline src/goto-programs/graphml_witness.cpp]: We can now also get rid of { and }

File: src/goto-programs/graphml_witness.cpp [hunk 3]
```diff
@@ -143,24 +170,35 @@ Function: graphml_witnesst::operator()
...
     if(it->hidden ||
        (!it->is_assignment() && !it->is_goto() && !it->is_assert()) ||
+       // we filter out steps with the same source location
+       // TODO: if these are assignments we should accumulate them into
+       //       a single edge
+       (prev_it!=goto_trace.steps.end() &&
+        prev_it->pc->source_location==it->pc->source_location) ||
+#if 0
        (it->is_goto() && it->pc->guard.is_true()) ||
+#endif
```
Comment (tautschnig) [inline src/goto-programs/graphml_witness.cpp]: Why do we no longer filter out goto?

File: src/xmllang/graphml.h [hunk 4]
```diff
@@ -32,9 +32,39 @@ struct xml_graph_nodet:public graph_nodet<xml_edget>
...
+  bool has_invariant;
+  std::string invariant;
+  std::string invariant_scope;
 };
 
-typedef graph<xml_graph_nodet> graphmlt;
+class graphmlt : public graph<xml_graph_nodet>
+{
+public:
+  inline bool has_node(std::string node_name) const
```
Comment (tautschnig) [inline src/xmllang/graphml.h]: Use `const std::string &node_name`

File: src/xmllang/graphml.h [hunk 5]
```diff
@@ -32,9 +32,39 @@ struct xml_graph_nodet:public graph_nodet<xml_edget>
...
+  {
+    for(node_indext i=0; i<nodes.size(); ++i)
+    {
+      if(nodes[i].node_name==node_name)
+        return true;
+    }
```
Comment (tautschnig) [inline src/xmllang/graphml.h]:
```
for(const auto & n : nodes)
  if(n.node_name==node_name)
    return true;
```

**Comment by tautschnig** [comment]
273ebde from #363 needs to be added and squashed to fix a bug introduced in dff110e of this series.

**Comment by peterschrammel** [comment]
@tautschnig, fixed.

**Comment by tautschnig** [comment]
@peterschrammel Would you mind squashing my latest commit into (my) previous one? I had made that mistake while using github's online editor... Then we should be good to go!

**Comment by peterschrammel** [comment]
Done.

**Pull request merged**